### PR TITLE
SAHI for small object detections

### PR DESCRIPTION
**Pull request by hardikdava**
# Description

This PR is implementation of #250 

## Type of change

Please delete options that are not relevant.

-   [x] New feature (non-breaking change which adds functionality)
-   [x] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

Google colab link:  https://colab.research.google.com/drive/1RNvsVWpkklUMZy5cjRs9y5yzBSHU9KRi?usp=sharing

## Docs

-   [x] Docs updated? What were the changes: `docs/supervision/detection/tools/slicer.md`

### CODE REVIEW

**Comment by hardikdava** [comment]
@SkalskiP, intial results. Object detection is working fine. I am trying to extend support for segmentation.

download

**Comment by SkalskiP** [comment]
Hi @hardikdava 👋🏻! Let's do object detection first. Merge it and go for segmentation than. 🙏🏻

**Comment by hardikdava** [comment]
model = "yolov8s.pt" (coco-pretrained model)
download (1)

**Comment by hardikdava** [comment]
@SkalskiP Please verify the results with bird image.

**Comment by SkalskiP** [comment]
@hardikdava result looks awesome! Curious why we missed that large bird. Any ideas?

257438537-fd8b7de4-a6e2-44bf-8a1e-7ae5346ea924

**Comment by SkalskiP** [comment]
@hardikdava I'm super excited. Do you want to write unit tests for this PR? Or like me to take care of it?

File: supervision/detection/tools/slicer.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        """
+        self.siced_width = sliced_width
+        self.sliced_height = sliced_height
+        self.overlap_width_ratio = overlap_width_ratio
+        self.overlap_height_ratio = overlap_height_ratio
+        self.iou_threshold = iou_threshold
+        self.callback = callback
+        _validate_callback(callback=callback)
+
+    def infer(self, image: np.ndarray) -> Detections:
```
[the remark this answers is not in the document]
Comment (hardikdava) [reply supervision/detection/tools/slicer.py:46]: Hahaha, I did it just like that.

File: supervision/detection/tools/slicer.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            (sv.Detections) repositioned detections result based on original image
+        """
+        if len(detection) == 0:
+            return detection
+        xyxy = Slicer._reposition_boxes(boxes=detection.xyxy, slice_location=slice_location)
+        detection.xyxy = xyxy
+        return detection
+
+    @staticmethod
+    def _reposition_boxes(boxes: np.ndarray, slice_location: Tuple[int, int, int, int]) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (hardikdava) [reply supervision/detection/tools/slicer.py:123]: @SkalskiP Do you think this API can be used somewhere else? Like with other feature? Then it make sense to expose for general purpose. I would suggest to limit it.

File: supervision/detection/tools/slicer.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        """
+        Args:
+            boxes (np.ndarray): boxes of model inference of the slice
+            slice_location (Tuple[int, int, int, int]): slice location at which inference was performed
+
+        Returns:
+            (np.ndarray) repositioned bounding boxes
+        """
+        boxes[:, 0] = boxes[:, 0] + slice_location[0]
```
[the remark this answers is not in the document]
Comment (hardikdava) [reply supervision/detection/tools/slicer.py:132]: @SkalskiP definately, we can do that.

File: supervision/detection/tools/slicer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (SkalskiP) [reply supervision/detection/tools/slicer.py:123]: @hardikdava Yup, I think `supervision.detection.utils` is the right home for this logic.

File: supervision/detection/tools/slicer.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            callback (Callable): model callback method which returns detections as sv.Detections
+            sliced_width (int): width of each slice
+            sliced_height (int): height of each slice
+            overlap_width_ratio (float): Fractional overlap in width of each
+                                        slice (e.g. an overlap of 0.2 for a slice of size 100 yields an
+                                        overlap of 20 pixels). Default 0.2.
+            overlap_height_ratio (float): Fractional overlap in height of each
+                                        slice (e.g. an overlap of 0.2 for a slice of size 100 yields an
+                                        overlap of 20 pixels). Default 0.2.
+            iou_threshold (float): non-maximum suppression iou threshold to remove overlapping detections
```
[the remark this answers is not in the document]
Comment (hardikdava) [reply supervision/detection/tools/slicer.py:36]: Currently, we are removing them with `nms`. But I was thinking about adding merging logic.

[the remark this answers is not in the document]
Comment (hardikdava) [reply supervision/detection/tools/slicer.py:36]: I will make an another PR for `weighted_merging_nms` for this.

**Comment by hardikdava** [comment]
@SkalskiP If you have tested it then we can merge it. There is no further work pending. I did all the reviews you requested.

**Comment by onuralpszr** [comment]
@hardikdava  would you also fix linter error too

**Comment by hardikdava** [comment]
@onuralpszr Done.

**Comment by SkalskiP** [comment]
@hardikdava looks like we will have the same `hardik` user problem on that PR as well.

**Comment by hardikdava** [comment]
@SkalskiP yeah, I was going to change that later today.

**Pull request closed without merging**